server/server.go

```diff
@@ -943,3 +948,68 @@ func isNotFound(err error) bool {
 
 	return false
 }
+
+// startReloadWatcherForMirrorRegistries starts a new goroutine that
+// monitors the "registries.conf.d" directory. If not set, it defaults to
+// "/etc/containers/registries.conf.d". The routine reloads the registries
+// configuration when a file is modified, created or deleted in the directory.
+func (s *Server) startReloadWatcherForMirrorRegistries(ctx context.Context, registriesConfDDir string) {
+	if registriesConfDDir == "" {
+		log.Infof(ctx, "No registries.conf.d directory specified, defaulting to /etc/containers/registries.conf.d")
+		registriesConfDDir = "/etc/containers/registries.conf.d"
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Fatalf(ctx, "Failed to create new watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	log.Infof(ctx, "Registered reload watcher for mirror registries configuration")
+
+	s.reloadMirrorRegistriesConfiguration(ctx, watcher, registriesConfDDir)
+}
+
+func (s *Server) reloadMirrorRegistriesConfiguration(ctx context.Context, watcher *fsnotify.Watcher, path string) {
+	if err := watcher.Add(path); err != nil {
+		log.Errorf(ctx, "Failed to add watcher for path %q: %s", path, err)
+		return
+	}
+
+	reloadLock := make(chan struct{}, 1)
+	reloadLock <- struct{}{}
+
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				return
+			}
+
+			// Check for .conf files only
+			if !strings.HasSuffix(filepath.Base(event.Name), ".conf") {
+				continue
+			}
+
+			<-reloadLock // Acquire the semaphore
+			go func() {
+				// Wait for some time before accepting more requests
+				time.Sleep(eventStaggerDuration)
```

Comment:
The net effect of this is still a bunch of routines queuing up to call ReloadRegistries. From what I can tell they'll all still enter this block (you only send one event on the channel). A WaitGroup may achieve what you're looking for here

Reply:
hm actually a waitgroup doesn't do it either...

Reply:
In that case, we need to rely on the rate limiter or the sliding window approach that I proposed earlier: https://github.com/cri-o/cri-o/pull/7788#discussion_r1513588252